### ISSUE TITLE
Question regarding config

### ISSUE THREAD

**Issue by lilynet336** [gkaindl/Community-mdnsResponder#3]
@gkaindl, I realize this project hasn't been updated since 2018, but I come to seek what is likely forbidden knowledge on Apple's networking software. Long story short, I want to setup a Wide Area Bonjour DNS server, and I need `dnsextd` for it. 

My question is, in the config file 
```
options {
	listen-on		port 53 {};
	nameserver		address 127.0.0.1 port 5030;
	private			port 5533;
	llq				port 5352;
};
```
What are `private` and `llq`? I understand that `listen-on` is where `dnsextd` should listen and `nameserver` is where it should go, but what are the other two? Do I need them in the config, do I need the ports publicly accessible, and do I need to map then via SRV records in any form? 

I cross my fingers you'll see this. Thank you so much.

**Comment by gkaindl** [comment]
`private` refers to the the port that dnsextd should use for queries using the [Private DNS](https://github.com/gkaindl/Community-mdnsResponder/blob/d36184f8f5df0a78996072ea4eee8da737dbf86d/mdns-patched/PrivateDNS.txt#L2) provided by wide-area bonjour – Unless you intend to use this, you can safely ignore it and keep it at the default.

`llq` is the port used for [long-lived queries](https://en.wikipedia.org/wiki/DNS_Long-Lived_Queries), which you can also leave at the default, unless you have a good reason to change it.

**Comment by lilynet336** [comment]
Thank you very much for replying! That makes sense now. Thanks for clarifying.

**Comment by lilynet336** [comment]
Well I tried to run it with `dnsextd` listening on port 5353 and forwarding to BIND on port 53. It seems as if `dnsextd` doesn't listen to the config file, because it says the address is already in use. The only thing that remedies this s making BIND use port 5030 and making `dnsextd` use port 53 (which I am leery about doing in case something happens to `dnsextd`, thereby causing clients to not be able to reach my DNS at all. Do you know if this was a reported bug, or if there's something I can do to fix this? I certainly do not expect you to try to patch it since it's so old, just wondering if there's a known solution.

**Comment by gkaindl** [comment]
Hmm, upon closer look, it appears that actually looking at these options is commented out (see [here](https://github.com/gkaindl/Community-mdnsResponder/blob/d36184f8f5df0a78996072ea4eee8da737dbf86d/mdns-patched/mDNSShared/dnsextd_parser.y#L154)), but that's not something I added (and probably didn't notice, because I never changed the default options, only the zone and key sections). The original code had all sort of these weird half-done implementation stubs and commented-out sections, and my patch was mostly about "filling in the blanks" to make it all run on non-OSX.

You could try un-commenting this line before building, but chances are that you'll get some build errors then, because of missing things. But maybe they aren't hard to fill in.

That being said, I've run dnsextd for my personal wide-area bonjour domain for many years now in the default configuration, e.g. on port 53, proxying to bind running on another port, and it's been super stable for me.

**Comment by lilynet336** [comment]
Well the whole half-baked thing does match with the rest of my experience with this stuff. Apple really never finished any of this... in fact, if you run `dns-sd -R` on macOS to register a service yourself, the manpage that comes with the tool says that domains other than `.local` will be supported "In the future". I will try removing the comment and see what happens, but if it's been stable for you then I will just see how it goes. Thank you again!

**Comment by lilynet336** [comment]
Sorry to reactivate this, but I still can't figure out, how do I create a zone key? `dnssec-keygen` says that HMAC-MD5 keys can't be used for zone signing, but host keys don't seem to work for zone updating. How did you deal with this for your personal zone?

EDIT: I have been using HOST instead of ZONE in the `dnssec-keygen` command, but getting `BADKEY` back from the server, so I assume I did something wrong. [dns-sd.org](http://dns-sd.org/ServerSetup.html) suggests using `dnssec-keygen -a HMAC-MD5 -b 128 -n ZONE <your zone>` which is where the error occurs. From research it seems like older versions of bind's libraries that would run this command just interpreted this as a mistake and made a HOST key anyways.

EDIT 2: The key has to be named after the domain in the `named.conf`. Oops... I feel like I originally had it like that, but it hasn't fully worked as of yet. Never mind my comment, I'll keep debugging. Sorry for the ping.

**Comment by lilynet336** [comment]
Actually, if you don't mind, can you share your configs? Specifically, the `key` declaration in `named.conf` and `dnsextd.conf` (redacted obviously), the `allow-update` declaration in `named.conf`, and then what to set clients to use for Domain, Username and Password under Dynamic Global Hostname? I used `nsupdate` with each individual TSIG key and BIND takes the updates, but macOS clients and my AirPort base stations can't seem to update. I am using the name of the key, like "bonjour.example.com" as the Domain and Username, and the TSIG secret as the password. Is this wrong? Sorry to yet again bother you with this, but I'm really interested in getting this working and I don't know anyone else to ask.

**Comment by lilynet336** [comment]
So I am still debugging. I have gotten to the point where BIND allows updates when I set `allow-update { 1.2.3.4; };`, but this is hardcoding the IPs which ruins the entire point of this setup. Something about the keys is just not right but I can't tell what. I have noticed that BIND occasionally `error: dumping master file: rename: /var/named/db.example.com: permission denied`, which is odd because I even tried `chmod a+rwx` the file temporarily and still get the error. The other weird part is, even when I `allow-update { any; };`, I can see the `adding an RR` lines in BIND's log, but when I try to browse for these records they don't exist at any point.

**Comment by gkaindl** [comment]
Just a quick update that I'll post my (redacted) config ASAP and look into the TSIG issue as well – sorry, couldn't squeeze it in so far!

**Comment by lilynet336** [comment]
No worries at all, take your time! Also I don't think the TSIG issue is with dnsextd, I think it's a named problem, specifically with how I've configured it.

**Comment by lilynet336** [comment]
It seems like modern BIND versions actually do LLQ (and maybe also UL) on their own. I have tried redoing my setup without `dnsextd` and I'm having a few weird (but probably unrelated) issues, and I think I'm going to try figuring out making `mdnsd` run on my servers and publish records that way. I *think* I was able to get Mac clients to work, and I got one of my AirPorts to work but the other just refuses (presumably a bug that will never be resolved).
